### ISSUE TITLE
Newbie Looking for Advice!

### ISSUE THREAD

**Issue by ashlyrjames** [RAHB-REALTORS-Association/constant-sync#8]
Hi Justin,

Sorry to add an "issue" to your project, but I was actually hoping to ask for some advice. This has been immensely helpful to me in trying to create an automation for a part time job! I struggled for so long with OAuth2, and after finding your repo it made complete sense. I'm currently hung up on that very issue though... I was just able to get my container running on Google Cloud Run Jobs but, if you're familiar with the service, there is no real CLI. You've already solved this problem I think based on glancing at the recent commits, but when running the code on my local computer previously I'd been manually going to Constant Contact's auth URL and manually pasting the callback URL to authenticate myself. This is an issue now that I'm trying to deploy to GCR. I'm missing an important piece of fundamental understanding when it comes to OAuth2 and how to initialize my access and set up recurring refresh tokens with CC API. Would you be able to help me? You can reach me at ashly.james@utexas.edu, I hope to hear from you!

Best,
Ashly

**Comment by justinh-rahb** [comment]
Hi Ashly,

I'm glad to hear that the project has been beneficial for you! I completely understand the challenges of automating the OAuth2 process, especially in deployment scenarios like Google Cloud Run. Let's dive into the details:

**Token Management**:
We utilize a dedicated JSON file (`tokens.json`) to store and manage access and refresh tokens.
```python
TOKEN_FILE = 'tokens.json'

def ensure_token_file_exists():
    if not os.path.exists(TOKEN_FILE):
        with open(TOKEN_FILE, "w") as file:
            json.dump({
                "access_token": "",
                "refresh_token": "",
                "expires_in": 0
            }, file)

# Call this function at the beginning to ensure tokens.json exists
ensure_token_file_exists()
```
This strategy is key for avoiding manual steps and ensuring an automated OAuth2 flow.

**OAuth2 Flow**:
1. **Refreshing the Access Token**:
```python
def refresh_access_token():
    token_data = load_tokens_from_file()
    refresh_token = token_data.get('refresh_token')
    
    if not refresh_token:
        return None
```
This function fetches the refresh token from the saved tokens file and, if present, uses it to request a new access token.

2. **Saving Tokens**:
```python
def save_tokens_to_file(access_token, refresh_token, expires_in):
    expiration_time = datetime.now() + timedelta(seconds=expires_in)
    token_data = {
        'access_token': access_token,
        'refresh_token': refresh_token,
        'expires_at': expiration_time.strftime('%Y-%m-%d %H:%M:%S')
    }
    with open(TOKEN_FILE, 'w') as f:
        json.dump(token_data, f)
```
After obtaining the tokens, they're persistently saved in the `tokens.json` file. This ensures the tokens are available for future requests and can be refreshed as necessary.

By managing tokens in this manner and understanding the OAuth2 flow, the need for manual intervention, such as visiting the auth URL, is minimized.

I hope this provides a clearer picture of the OAuth2 process implemented in the repository. Please don't hesitate to reach out if you have further questions or need more insights!

Best wishes,
Justin

**Comment by ashlyrjames** [comment]
Hello! 

It's taken me awhile but I'm implementing your above advice and working to adapt your repo to my project! I've hit a wall related to oauth2 yet again though... I feel like it's something simple but cannot figure out what. I have created a tokens.json file but it's currently empty of any data. I have added the same get_access_token() function:
`def get_access_token():
    token_data = load_tokens_from_file()
    if not token_data:
        return None
    
    if is_token_expired(token_data):
        return refresh_access_token()
    else:
        return token_data.get('access_token')`
 but my code errors when trying to GET/POST to ConstantContact. Specifically when I call the following function:

`def create_or_update_contact(self, data):
        url = self.api_url + '/contacts/sign_up_form'
        headers = {
            'Authorization': 'Bearer ' + self.access_token,
            'Content-Type': 'application/json'
        }
        response = requests.post(url, headers=headers, json=data)
        response.raise_for_status()
        return response.json() `

I'm getting the error "TypeError: can only concatenate str (not "NoneType") to str" at the following line:
`'Authorization': 'Bearer ' + self.access_token, `

My code is an amalgamation of a previous version of constant-sync that utilized Classes (hence the self.access_token) so it may be that I'm trying to Frankenstein parts together in a way that simply won't work... Here's the init method for a little more context:

class ConstantContactAPI:
    def __init__(self, access_token):
        self.access_token = get_access_token()
        self.api_url = 'https://api.cc.email/v3'`

In short, did I miss something simple or is there a larger disconnect in my code generally? I appreciate your help! It's been difficult to adapt my older version to the recent commits for constant-sync, so if the solution is follow along more closely to match my project to yours (i.e. drop the classes, implement the "HEADERS" and get_auth_headers() in api_integration.py) I'm happy to do so! If I've left out any information or need to clarify anything please ask, I appreciate your help so much! 

Best,

Ashly

**Comment by justinh-rahb** [comment]
Hello Ashly,

That's fantastic news—congratulations on successfully implementing the changes and overcoming the OAuth2 hurdles! It's always rewarding to see hard work and persistence pay off.

Interestingly, your journey mirrors my own experience with this project. I actually had to rewrite the entire codebase from scratch to solve more or less the same problems you were facing, especially with OAuth2. It's reassuring to hear that you got it to work, as it validates the effectiveness of the latest changes. 

If you have any more questions or run into other challenges, don't hesitate to reach out. Otherwise, best of luck with the rest of your project!

Warm regards,
Justin
